hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    on_ble_scan_result_cb_t scanResultCallBack_;            /**< Callback function on scan result. */
+    void* context_;                                         /**< Context of the scan result callback function. */
+
+    static const uint16_t OBSERVER_SCANNED_DATA_POOL_SIZE = 1024;
+
+    bool chachedDevice(const hal_ble_addr_t& address) const;
+    hal_ble_gap_on_scan_result_evt_t* chachedResult(const hal_ble_addr_t& address);
+    void removeResult(const hal_ble_addr_t& address);
+    void clearResult(void);
+    int continueScanning(void);
+    static void observerEventProcessedHook(const hal_ble_evts_t *event, void* context);
+    static void processObserverEvents(const ble_evt_t* event, void* context);
 };
 
-struct Conn {
-    uint16_t handle;
-    uint16_t paramUpdateAttempts;
+class BleObject::ConnectionsManager {
+public:
+    typedef struct {
+        uint8_t role;
+        uint16_t connHandle;
+        hal_ble_conn_params_t effectiveConnParams;
+        hal_ble_addr_t peer;
+    } BleConnection_t;
+
+    ConnectionsManager(BleObject* obj);
+    ~ConnectionsManager() = default;
+    int init(void);
+    int setPPCP(const hal_ble_conn_params_t* ppcp);
+    int getPPCP(hal_ble_conn_params_t* ppcp) const;
+    bool connecting(void) const;
+    bool connected(const hal_ble_addr_t* address=nullptr);
+    int connect(const hal_ble_addr_t* address);
+    int connectCancel(void);
+    int disconnect(uint16_t connHandle);
+    int updateConnectionParams(uint16_t connHandle, const hal_ble_conn_params_t* params);
+    int getConnectionParams(uint16_t connHandle, hal_ble_conn_params_t* params);
+    bool valid(uint16_t connHandle);
+
+private:
+    uint8_t isConnecting_ : 1;                      /**< If it is connecting or not. */
+    hal_ble_addr_t connectingAddr_;                 /**< Address of peer the Central is connecting to. */
+    uint16_t disconnectingHandle_;                  /**< Handle of connection to be disconnected. */
+    uint16_t connParamUpdateHandle_;                /**< Handle of the connection that is to send peripheral connection update request. */
+    uint8_t connParamsUpdateAttempts_;              /**< Attempts for peripheral to update connection parameters. */
+    os_timer_t connParamsUpdateTimer_;              /**< Timer used for sending peripheral connection update request after connection established. */
+    os_semaphore_t connParamsUpdateSemaphore_;      /**< Semaphore to wait until connection parameters updated. */
+    os_semaphore_t connectSemaphore_;               /**< Semaphore to wait until connection established. */
+    os_semaphore_t disconnectSemaphore_;            /**< Semaphore to wait until connection disconnected. */
+    Vector<BleConnection_t> connections_;           /**< Current on-going connections. */
+    BleObject* bleObj_;                             /**< Pointer to the BLE Object. */
+
+    static const uint32_t CONNECTION_OPERATION_TIMEOUT_MS = 5000;
```

Comment:
Consider moving this out of class into the anonymous namespace in the top of the source file for better visibility.